File: 프로그래머스/3/12907. 거스름돈/거스름돈.java
class Solution {
    public int solution(int n, int[] money) {
        // dp[i]는 i원을 거슬러 줄 수 있는 경우의 수를 저장
        int[] dp = new int[n + 1];
        dp[0] = 1; // 0원을 거슬러 주는 방법은 1가지 (아무 동전도 사용하지 않음)

        // 각 동전 단위에 대해 경우의 수를 계산
        for (int coin : money) {
            for (int i = coin; i <= n; i++) {
                dp[i] = (dp[i] + dp[i - coin]) % 1000000007;
            }
        }

        // n원을 거슬러 줄 수 있는 경우의 수 반환
        return dp[n];
    }
}

File: 프로그래머스/3/12907. 거스름돈/거스름돈.py
def solution(n, money):
    # DP 배열 초기화
    dp = [0] * (n + 1)
    dp[0] = 1  # 0원을 만드는 방법은 1가지
    
    # 동전 순회
    for coin in money:
        for i in range(coin, n + 1):
            dp[i] += dp[i- coin]
            dp[i] %= 1_000_000_007  # 나머지 연산 적용
            
    return dp[n]